### PR TITLE
Download many files in one go

### PR DESCRIPTION
**Pull request by oscargus**
Implements #1496 and also allows selecting multiple documents to find full text documents for. I think that it doesn't always show a dialog for every failed files when having multiple ongoing downloads.

Please try it out at http://builds.jabref.org/downloadmanyfiles

- [x] Change in CHANGELOG.md described
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

**Comment by Siedlerchr** [comment]
If it does what you promise, I love you! :heart_eyes:  :laughing:  Will try this tomorrow. 
>think that it doesn't always show a dialog for every failed files when having multiple ongoing downloads.

Not sure if I understand you correctly: So I download e.g. 20 files, and 5 fails. Is then the dialog "Failed DL" displayed 5 times?
Or after all downloads are finished it shows a dialog with e.g. 15/20 files downloaded succesfully?

File: src/main/java/net/sf/jabref/external/FindFullTextAction.java [hunk 1]
```diff
@@ -41,52 +44,60 @@ public void init() throws Throwable {
...
+                List<String> dirs = basePanel.getBibDatabaseContext()
+                        .getFileDirectory(Globals.prefs.getFileDirectoryPreferences());
+                if (dirs.isEmpty()) {
+                    JOptionPane.showMessageDialog(basePanel.frame(),
+                            Localization.lang("Main file directory not set!") + " " + Localization.lang("Preferences")
+                                    + " -> " + Localization.lang("External programs"),
+                            Localization.lang("Directory not found"), JOptionPane.ERROR_MESSAGE);
+                    return;
+                }
+                // TODO: this needs its own thread as it blocks the UI!
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/external/FindFullTextAction.java]: Is this blocking still a problem?

**Comment by Siedlerchr** [comment]
Works technically fine, but I have some remarks, as it's currently not effective usable:
I ran it over Quality->Look up fulltext
When the process is finished, it shows the file save/ Download failed dialog for every entry which I previously selected.
So if I have 200 entries select, after the process it opens 200 dialogs...
I would suggest having just a final download result dialog or directly a process window, perhaps sth like this from Down Them All:
dta


There is apparently no option to cancel the lookup process.

File: src/main/java/net/sf/jabref/external/FindFullTextAction.java (hunk 1 above)
Comment (oscargus) [inline src/main/java/net/sf/jabref/external/FindFullTextAction.java]: Not sure if it was before either...

**Comment by oscargus** [comment]
Well, what you select is what you get. :-)

I agree that your approach would be even better, but that is out of the scope for me at the moment. One could still reenable one selected entry at a time, with the added benefit of being able to start more than one, but I do not really know if that is a better way to do it. Maybe it is better to pop up a dialog asking if the user is sure they want to try to find > 10(?) full text documents as this can be quite a hassle?

**Comment by Siedlerchr** [comment]
Agree with you. Add this dialog warning, for more than x? entries .
For the rest we should create a new issue/enhancement. Mabye the other devs have some other 

Regarding the blocking: Does not block the UI atm.

**Comment by oscargus** [comment]
I updated and set it to 10. Not sure how much one should actually write in the dialog...

File: src/main/java/net/sf/jabref/external/FindFullTextAction.java [hunk 2]
```diff
@@ -44,6 +44,20 @@ public void init() throws Throwable {
 
     @Override
     public void run() {
+        if (basePanel.getSelectedEntries().size() >= 10) {
```
Comment (Siedlerchr) [inline src/main/java/net/sf/jabref/external/FindFullTextAction.java]: Please extract that "magic number" to a constant.
https://en.wikipedia.org/wiki/Magic_number_(programming)

**Comment by Siedlerchr** [comment]
Code LGTM, just move that magic number to a constant.

**Comment by oscargus** [comment]
Yeah, was thinking about the constant thing as well. Will fix it. I guess it is a good idea to merge now so people have time to find possible issues.

**Comment by oscargus** [comment]
Improved the file name generation as well. If the formatter does return an empty string, the BibTeX key is used. If no BibTeX key, "default" is used.

**Pull request merged**